### ISSUE TITLE
[Poison] Add detection of reference assemblies

### ISSUE THREAD

**Issue by crummel** [dotnet/source-build#2817]
Currently, the leak detection infrastructure only poisons the previously-source-built nupkgs.  We've talked with the runtime team and we also shouldn't ever be shipping reference assemblies - we can augment the poison reporting to detect these to help with this goal:
- MarkAndCatalog task should include the ref packages to record the hashes of those assemblies as well.
- CheckForPoison should include a new `PoisonType`, `ReferenceAttribute` or similar, that checks for the `ReferenceAssembly` attribute on shipped assemblies.

**Comment by ellahathaway** [comment]
In discussing this issue with @NikolaMilosavljevic, it seems that adding in this feature is something new. Currently, the reference assemblies are coming from the SBRP repo which is built after we poison the files. As such, we've never done poisoning steps during build, which addressing this issue would require us to do.

This might warrant some brainstorming.

**Comment by ellahathaway** [comment]
In trying to address this, I'm having trouble finding some of the .dll files that are present in the SDK.

One reference assembly in particular is Microsoft.Win32.SystemEvents.dll. I expected this assembly to be in a nuget package or as a plain binary in the dotnet repo, but no such luck. The best I can trace it to is "./installer/src/SourceBuild/content/test/Microsoft.DotNet.SourceBuild.SmokeTests/bin/Debug/net8.0/.dotnet/sdk/9.0.100-alpha.1.23431.1/Microsoft.Win32.SystemEvents.dll" and "./dotnet/src/installer/artifacts/source-build/self/src/artifacts/bin/redist/Release/dotnet/sdk/9.0.100-alpha.1.23422.1/Microsoft.Win32.SystemEvents.dll". Is there an explanation for why this assembly is not present in PSB or SBRP?

**Comment by mthalman** [comment]
It is in SBRP: https://github.com/dotnet/source-build-reference-packages/tree/main/src/referencePackages/src/microsoft.win32.systemevents/7.0.0. It would be in `/vmr/prereqs/packages/reference` during the build.

**Comment by ellahathaway** [comment]
I should've been more specific. When poisoning the SBRP (which included `/vmr/prereqs/packages/reference` and `/vmr/artifacts/obj/x64/Release/blob-feed/packages/SourceBuildReferencePackages`, done in different builds), I found that the reference assembly would be in the poison catalog, but the assembly was still in the resulting sdk and was not in the poison usage report. This led me to realize that the sha256 hashes of the ref assemblies in the catalog, all with the same name as the assembly in the sdk, did not match the hash of the assembly in the sdk. Additionally, the assembly in the sdk was missing the poison marker, indicating that it had not been poisoned and cataloged.

I've found multiple instances of the reference assembly with the same name in the repo, but none have the same sha258 hash and binary code as the reference assembly in the sdk. [Here is a report ](https://gist.github.com/ellahathaway/650f059100e9eeee6ad8dc4e1641db66#file-dotnet_prereqs_dllsearch-txt)of the sha256 hash of the reference assembly in the sdk and the other hashes of matching names in `/vmr/prereqs/`
